File: runners/go/interpret.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package main

import (
	"context"
	"fmt"
	"path/filepath"
	"strings"

	"github.com/apache/iceberg-go"
	"github.com/apache/iceberg-go/catalog"
	iceio "github.com/apache/iceberg-go/io"
	icetable "github.com/apache/iceberg-go/table"
)

// unsupportedError signals an op/kind this implementation does not support
// (declared gap). main maps it to exit code 4.
type unsupportedError struct {
	entryIdx int
	feature  string
}

func (e *unsupportedError) Error() string {
	return fmt.Sprintf("entry %d: unsupported feature %q", e.entryIdx, e.feature)
}

// runner executes an L-log against iceberg-go and accumulates canonical output.
type runner struct {
	ctx      context.Context
	cat      catalog.Catalog
	ident    icetable.Identifier
	tbl      *icetable.Table
	specPath string // path to the l-log file (to resolve fixture-relative artifact dirs)

	schema   *iceberg.Schema
	canonIDs map[string]int // column name -> canonical output field-id (__rowkey => 0)

	// snapshotOrdinal maps an iceberg snapshot-id to its commit ordinal
	// (0,1,2...) so output references the ordinal, not the random id.
	snapshotOrdinal map[int64]int
	nextOrdinal     int

	// binds maps an observe 'bind' name to the snapshot-id current at that
	// point, so a later 'observe at: <name>' time-travels to it.
	binds map[string]int64

	out *CanonicalOutput
}

// execute acquires the table (synthesized from the header, or loaded from a
// checked-in artifact) then processes each entry in order, assembling canonical
// output.
func (r *runner) execute(log *LLog) error {
	r.snapshotOrdinal = map[int64]int{}
	r.binds = map[string]int64{}
	r.out = &CanonicalOutput{SpecID: log.Header.ID, Accept: true}

	if log.Header.Source == "artifact" {
		if err := r.loadArtifact(log.Header); err != nil {
			return fmt.Errorf("load artifact: %w", err)
		}
	} else {
		sc, _, err := buildSchema(log.Header.Schema)
		if err != nil {
			return fmt.Errorf("build schema: %w", err)
		}
		r.schema = sc
		r.canonIDs = buildCanonIDs(log.Header.Schema)
		if err := r.createTable(log.Header); err != nil {
			return fmt.Errorf("create table: %w", err)
		}
	}

	for i, e := range log.Entries {
		if err := r.applyEntry(i, e); err != nil {
			return err
		}
	}
	return nil
}

// loadArtifact materializes a checked-in table (restoring its bytes to the
// pinned root so embedded absolute paths resolve) and loads it read-only. The
// canonical field-ids are derived by name from the loaded schema; __rowkey => 0
// and user columns keep the field-id the table carries (same relabel rule the
// synthesized path uses).
func (r *runner) loadArtifact(h Header) error {
	if h.Artifact == nil {
		return fmt.Errorf("source: artifact requires an 'artifact' block")
	}
	fixtureDir := filepath.Join(filepath.Dir(r.specPath), h.Artifact.Path)
	metaPath, err := materialize(fixtureDir)
	if err != nil {
		return err
	}
	tbl, err := icetable.NewFromLocation(
		r.ctx, icetable.Identifier{"default", "t"}, metaPath,
		iceio.LoadFSFunc(nil, metaPath), nil,
	)
	if err != nil {
		return fmt.Errorf("NewFromLocation %s: %w", metaPath, err)
	}
	r.tbl = tbl
	r.canonIDs = canonIDsFromSchema(tbl.Schema())
	return nil
}

// createTable creates the namespace + table from the header schema and
// properties.
func (r *runner) createTable(h Header) error {
	ns := icetable.Identifier{"default"}
	if err := r.cat.CreateNamespace(r.ctx, ns, nil); err != nil {
		return err
	}
	opts := []catalog.CreateTableOpt{}
	props := iceberg.Properties{
		"format-version": fmt.Sprintf("%d", h.FormatVersion),
	}
	for k, v := range h.Properties {
		props[k] = v
	}
	opts = append(opts, catalog.WithProperties(props))

	tbl, err := r.cat.CreateTable(r.ctx, r.ident, r.schema, opts...)
	if err != nil {
		return err
	}
	r.tbl = tbl
	return nil
}

// applyEntry dispatches one op-log entry.
func (r *runner) applyEntry(idx int, e Entry) error {
	switch e.Op {
	case "append":
		return r.doAppend(idx, e)
	case "observe":
		return r.doObserve(idx, e)
	case "delete":
		return r.doDelete(idx, e)
	default:
		return &unsupportedError{entryIdx: idx, feature: "op." + e.Op}
	}
}

// doAppend builds an arrow table from the rows and commits one append snapshot.
func (r *runner) doAppend(idx int, e Entry) error {
	arrowTbl, err := buildArrowTable(r.tbl.Schema(), e.Rows)
	if err != nil {
		return fmt.Errorf("entry %d append: %w", idx, err)
	}
	defer arrowTbl.Release()

	tx := r.tbl.NewTransaction()
	if err := tx.AppendTable(r.ctx, arrowTbl, arrowTbl.NumRows(), nil); err != nil {
		return fmt.Errorf("entry %d append: %w", idx, err)
	}
	tbl, err := tx.Commit(r.ctx)
	if err != nil {
		return fmt.Errorf("entry %d commit: %w", idx, err)
	}
	r.tbl = tbl
	return r.recordSnapshot()
}

// doDelete performs a delete. The spec-open strategy is mapped to the
// write.delete.mode table property before Transaction.Delete: merge-on-read
// writes position deletes (content=1); copy-on-write rewrites data files.
// Equality deletes and deletion vectors are declared gaps (exit 4).
func (r *runner) doDelete(idx int, e Entry) error {
	if e.Kind == "equality" {
		return &unsupportedError{entryIdx: idx, feature: "delete.merge-on-read.equality"}
	}
	if e.Kind == "deletion-vector" {
		return &unsupportedError{entryIdx: idx, feature: "delete.merge-on-read.deletion-vector"}
	}

	var mode string
	switch e.Strategy {
	case "merge-on-read":
		mode = icetable.WriteModeMergeOnRead
	case "copy-on-write":
		mode = icetable.WriteModeCopyOnWrite
	default:
		return fmt.Errorf("entry %d delete: unknown strategy %q", idx, e.Strategy)
	}

	if e.Predicate == nil {
		return fmt.Errorf("entry %d delete: missing predicate", idx)
	}
	filter, err := r.buildExpr(e.Predicate)
	if err != nil {
		return fmt.Errorf("entry %d delete: %w", idx, err)
	}

	tx := r.tbl.NewTransaction()
	if err := tx.SetProperties(iceberg.Properties{icetable.WriteDeleteModeKey: mode}); err != nil {
		return fmt.Errorf("entry %d set delete mode: %w", idx, err)
	}
	if err := tx.Delete(r.ctx, filter, nil); err != nil {
		return fmt.Errorf("entry %d delete: %w", idx, err)
	}
	tbl, err := tx.Commit(r.ctx)
	if err != nil {
		return fmt.Errorf("entry %d commit: %w", idx, err)
	}
	r.tbl = tbl
	return r.recordSnapshot()
}

// buildExpr converts an L-log predicate to an iceberg BooleanExpression. The
// term's column type drives literal parsing.
func (r *runner) buildExpr(p *Predicate) (iceberg.BooleanExpression, error) {
	switch p.Type {
	case "true":
		return iceberg.AlwaysTrue{}, nil
	case "false":
		return iceberg.AlwaysFalse{}, nil
	case "is-null":
		return iceberg.UnaryPredicate(iceberg.OpIsNull, iceberg.Reference(p.Term)), nil
	case "not-null":
		return iceberg.UnaryPredicate(iceberg.OpNotNull, iceberg.Reference(p.Term)), nil
	case "eq", "ne", "lt", "lt-eq", "gt", "gt-eq":
		lit, err := r.termLiteral(p.Term, p.Value)
		if err != nil {
			return nil, err
		}
		return iceberg.LiteralPredicate(cmpOp(p.Type), iceberg.Reference(p.Term), lit), nil
	case "in", "not-in":
		lits := make([]iceberg.Literal, 0, len(p.Values))
		for _, v := range p.Values {
			lit, err := r.termLiteral(p.Term, v)
			if err != nil {
				return nil, err
			}
			lits = append(lits, lit)
		}
		op := iceberg.OpIn
		if p.Type == "not-in" {
			op = iceberg.OpNotIn
		}
		return iceberg.SetPredicate(op, iceberg.Reference(p.Term), lits), nil
	case "and", "or":
		if len(p.Args) < 2 {
			return nil, fmt.Errorf("%s needs >= 2 args", p.Type)
		}
		sub := make([]iceberg.BooleanExpression, 0, len(p.Args))
		for _, a := range p.Args {
			e, err := r.buildExpr(a)
			if err != nil {
				return nil, err
			}
			sub = append(sub, e)
		}
		if p.Type == "and" {
			return iceberg.NewAnd(sub[0], sub[1], sub[2:]...), nil
		}
		return iceberg.NewOr(sub[0], sub[1], sub[2:]...), nil
	case "not":
		if p.Arg == nil {
			return nil, fmt.Errorf("not needs an arg")
		}
		e, err := r.buildExpr(p.Arg)
		if err != nil {
			return nil, err
		}
		return iceberg.NewNot(e), nil
	default:
		return nil, fmt.Errorf("unsupported predicate type %q", p.Type)
	}
}

// termLiteral parses a typed value against the column type named by term.
func (r *runner) termLiteral(term string, tv *TypedValue) (iceberg.Literal, error) {
	if tv == nil {
		return nil, fmt.Errorf("predicate on %q missing value", term)
	}
	typ, ok := r.tbl.Schema().FindTypeByName(term)
	if !ok {
		return nil, fmt.Errorf("predicate term %q not in schema", term)
	}
	return tv.toLiteral(typ)
}

// cmpOp maps an l-log comparison predicate type to an iceberg Operation.
func cmpOp(t string) iceberg.Operation {
	switch t {
	case "eq":
		return iceberg.OpEQ
	case "ne":
		return iceberg.OpNEQ
	case "lt":
		return iceberg.OpLT
	case "lt-eq":
		return iceberg.OpLTEQ
	case "gt":
		return iceberg.OpGT
	case "gt-eq":
		return iceberg.OpGTEQ
	}
	return iceberg.OpEQ
}

// doObserve scans the table at the requested point and appends an observation.
// Phase 0/1 only handle 'latest'; time-travel to a bound name is Phase 2.
func (r *runner) doObserve(idx int, e Entry) error {
	at, err := observeAt(e)
	if err != nil {
		return fmt.Errorf("entry %d observe: %w", idx, err)
	}

	var scan *icetable.Scan
	switch at.kind {
	case atLatest:
		scan = r.tbl.Scan()
	case atBind:
		snapID, ok := r.binds[at.bind]
		if !ok {
			return fmt.Errorf("entry %d observe: unknown bind %q", idx, at.bind)
		}
		scan = r.tbl.Scan(icetable.WithSnapshotID(snapID))
	case atOrdinal:
		snapID, ok := r.ordinalSnapshot(at.ordinal)
		if !ok {
			return fmt.Errorf("entry %d observe: unknown ordinal %d", idx, at.ordinal)
		}
		scan = r.tbl.Scan(icetable.WithSnapshotID(snapID))
	default:
		return &unsupportedError{entryIdx: idx, feature: "time-travel"}
	}

	arrowTbl, err := scan.ToArrowTable(r.ctx)
	if err != nil {
		return fmt.Errorf("entry %d scan: %w", idx, err)
	}
	defer arrowTbl.Release()

	fields := r.tbl.Schema().Fields()
	icebergSchema, rows, err := decodeScan(arrowTbl, fields, r.canonIDs)
	if err != nil {
		return fmt.Errorf("entry %d decode: %w", idx, err)
	}

	// 'at' echoes the resolved observation point: when the observe binds a
	// name, that name is the resolved point (the golden reports the bind name,
	// not the literal 'latest' the author wrote); otherwise the literal target.
	atValue := at.value
	if e.Bind != "" {
		atValue = e.Bind
	}

	obs := Observation{
		At:            atValue,
		Bind:          e.Bind,
		IcebergSchema: icebergSchema,
		DecodedRows:   rows,
	}
	r.out.Observations = append(r.out.Observations, obs)

	// record the bind -> current snapshot for later time-travel
	if e.Bind != "" {
		if snap := r.tbl.CurrentSnapshot(); snap != nil {
			r.binds[e.Bind] = snap.SnapshotID
		}
	}
	return nil
}

// recordSnapshot assigns the newest snapshot its commit ordinal and appends a
// SnapshotOut with the spec-pinned facts.
func (r *runner) recordSnapshot() error {
	snap := r.tbl.CurrentSnapshot()
	if snap == nil {
		return fmt.Errorf("no current snapshot after commit")
	}
	if _, seen := r.snapshotOrdinal[snap.SnapshotID]; seen {
		return nil
	}
	ordinal := r.nextOrdinal
	r.snapshotOrdinal[snap.SnapshotID] = ordinal
	r.nextOrdinal++

	so := SnapshotOut{Ordinal: ordinal}
	if snap.ParentSnapshotID != nil {
		if p, ok := r.snapshotOrdinal[*snap.ParentSnapshotID]; ok {
			so.Parent = &p
		}
	}
	if snap.Summary != nil {
		so.Operation = string(snap.Summary.Operation)
		so.Summary = summaryOut(snap.Summary.Properties)
	}

	deleteFiles, err := r.snapshotDeleteFiles(snap)
	if err != nil {
		return fmt.Errorf("collect delete files: %w", err)
	}
	so.DeleteFiles = deleteFiles

	r.out.Snapshots = append(r.out.Snapshots, so)
	return nil
}

// ordinalSnapshot reverse-looks-up the snapshot-id for a commit ordinal.
func (r *runner) ordinalSnapshot(ordinal int) (int64, bool) {
	for id, ord := range r.snapshotOrdinal {
		if ord == ordinal {
			return id, true
		}
	}
	return 0, false
}

// snapshotDeleteFiles walks the delete manifests added by this snapshot and
// emits the spec-pinned facts (content, format, equality-ids) for each delete
// file — no paths, sizes, or counts. Only delete files added in this snapshot
// (matching its sequence number) are reported.
func (r *runner) snapshotDeleteFiles(snap *icetable.Snapshot) ([]DeleteFile, error) {
	fs, err := r.tbl.FS(r.ctx)
	if err != nil {
		return nil, err
	}
	manifests, err := snap.Manifests(fs)
	if err != nil {
		return nil, err
	}

	var out []DeleteFile
	for _, m := range manifests {
		if m.ManifestContent() != iceberg.ManifestContentDeletes {
			continue
		}
		// Only delete manifests added by this snapshot (the manifest's
		// added-snapshot-id is reliably set; per-entry snapshot ids may be
		// inherited/nil for freshly written files).
		if m.SnapshotID() != snap.SnapshotID {
			continue
		}
		entries, err := m.FetchEntries(fs, true) // discardDeleted: only live delete files
		if err != nil {
			return nil, err
		}
		for _, entry := range entries {
			df := entry.DataFile()
			out = append(out, DeleteFile{
				Content: int(df.ContentType()),
				Format:  fileFormatName(df.FileFormat()),
			})
		}
	}
	return out, nil
}

// fileFormatName lowercases iceberg's uppercase FileFormat enum to match the
// golden vocabulary ("parquet", "puffin").
func fileFormatName(f iceberg.FileFormat) string {
	return strings.ToLower(string(f))
}

File: runners/rust/src/interpret.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

//! The runner state machine: materialize the table from the header, then apply
//! each op-log entry, assembling canonical output. Emit-only.
//!
//! Capability note (iceberg-rust HEAD): append + scan + time-travel are
//! supported. Merge-on-read delete WRITE, schema type-promotion, and
//! compaction are not committable via any public API, so those ops raise
//! [`RunError::Unsupported`] (mapped to exit 4 — a declared gap in the matrix).

use std::collections::HashMap;

use anyhow::{Result, anyhow, bail};
use arrow_array::RecordBatch;
use futures::TryStreamExt;
use iceberg::spec::{DataFileFormat, FormatVersion};
use iceberg::transaction::{ApplyTransactionAction, Transaction};
use iceberg::writer::base_writer::data_file_writer::DataFileWriterBuilder;
use iceberg::writer::file_writer::ParquetWriterBuilder;
use iceberg::writer::file_writer::location_generator::{
    DefaultFileNameGenerator, DefaultLocationGenerator,
};
use iceberg::writer::file_writer::rolling_writer::RollingFileWriterBuilder;
use iceberg::writer::{IcebergWriter, IcebergWriterBuilder};
use iceberg::{Catalog, NamespaceIdent, TableCreation};
use iceberg_catalog_sql::SqlCatalog;
use parquet::file::properties::WriterProperties;
use serde_yaml::Value as YamlValue;

use crate::arrow_build::{Row, build_record_batch};
use crate::arrow_decode::decode_scan;
use crate::emit::{CanonicalOutput, Observation, SnapshotOut, SummaryOut};
use crate::llog::{Entry, LLog};
use crate::schema::{ROW_KEY_NAME, build_canon_ids, build_schema};
use crate::values::TypedValue;

/// An error carrying its category so main can map it to the right exit code.
#[derive(Debug)]
pub enum RunError {
    /// Op/kind unsupported by iceberg-rust (declared gap) -> exit 4.
    Unsupported { entry: usize, feature: String },
    /// Any other execution failure -> exit 1.
    Other(anyhow::Error),
}

impl std::fmt::Display for RunError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            RunError::Unsupported { entry, feature } => {
                write!(f, "entry {entry}: unsupported feature {feature:?}")
            }
            RunError::Other(e) => write!(f, "{e}"),
        }
    }
}

impl From<anyhow::Error> for RunError {
    fn from(e: anyhow::Error) -> Self {
        RunError::Other(e)
    }
}

/// Where an observe reads from.
enum At {
    Latest,
    Bind(String),
    Ordinal(i64),
}

pub struct Runner {
    catalog: SqlCatalog,
    table: iceberg::table::Table,
    canon_ids: HashMap<String, i32>,
    /// iceberg snapshot-id -> commit ordinal (0,1,2...).
    snapshot_ordinal: HashMap<i64, i64>,
    next_ordinal: i64,
    /// bind name -> snapshot-id current at that observe.
    binds: HashMap<String, i64>,
    out: CanonicalOutput,
}

impl Runner {
    /// Create the namespace + table from the header, returning a ready Runner.
    pub async fn materialize(catalog: SqlCatalog, log: &LLog) -> Result<Runner, RunError> {
        let schema = build_schema(&log.header.schema)?;
        let canon_ids = build_canon_ids(&log.header.schema);

        let ns = NamespaceIdent::new("default".to_string());
        catalog
            .create_namespace(&ns, HashMap::new())
            .await
            .map_err(|e| RunError::Other(anyhow!("create namespace: {e}")))?;

        let format_version = match log.header.format_version {
            1 => FormatVersion::V1,
            3 => FormatVersion::V3,
            _ => FormatVersion::V2,
        };

        let creation = TableCreation::builder()
            .name("t".to_string())
            .schema(schema)
            .format_version(format_version)
            .properties(log.header.properties.clone())
            .build();
        let table = catalog
            .create_table(&ns, creation)
            .await
            .map_err(|e| RunError::Other(anyhow!("create table: {e}")))?;

        Ok(Runner {
            catalog,
            table,
            canon_ids,
            snapshot_ordinal: HashMap::new(),
            next_ordinal: 0,
            binds: HashMap::new(),
            out: CanonicalOutput {
                spec_id: log.header.id.clone(),
                accept: true,
                snapshots: Vec::new(),
                observations: Vec::new(),
            },
        })
    }

    pub async fn run(&mut self, log: &LLog) -> Result<(), RunError> {
        for (idx, entry) in log.entries.iter().enumerate() {
            self.apply_entry(idx, entry).await?;
        }
        Ok(())
    }

    pub fn into_output(self) -> CanonicalOutput {
        self.out
    }

    async fn apply_entry(&mut self, idx: usize, e: &Entry) -> Result<(), RunError> {
        match e.op.as_str() {
            "append" => self.do_append(idx, e).await,
            "observe" => self.do_observe(idx, e).await,
            // Declared gaps: iceberg-rust has no committable MoR delete write
            // path, no type-promotion, and no compaction action.
            "delete" => Err(RunError::Unsupported {
                entry: idx,
                feature: delete_feature(e),
            }),
            "overwrite" => Err(RunError::Unsupported {
                entry: idx,
                feature: "op.overwrite".to_string(),
            }),
            "rewrite" => Err(RunError::Unsupported {
                entry: idx,
                feature: "rewrite".to_string(),
            }),
            "evolve-schema" => Err(RunError::Unsupported {
                entry: idx,
                feature: "evolve-schema.promote-type".to_string(),
            }),
            "evolve-spec" => Err(RunError::Unsupported {
                entry: idx,
                feature: "evolve-spec".to_string(),
            }),
            other => Err(RunError::Unsupported {
                entry: idx,
                feature: format!("op.{other}"),
            }),
        }
    }

    async fn do_append(&mut self, idx: usize, e: &Entry) -> Result<(), RunError> {
        let rows = parse_rows(e)
            .map_err(|err| RunError::Other(err.context(format!("entry {idx} append"))))?;
        let batch = build_record_batch(self.table.metadata().current_schema(), &rows)
            .map_err(|err| RunError::Other(err.context(format!("entry {idx} append"))))?;

        // Unique file-name prefix per append: DefaultFileNameGenerator resets
        // its counter to 00000 for each new writer, so a fixed prefix would
        // collide across appends and fast_append's duplicate check would reject
        // the second file. Keying on the entry index keeps every file distinct.
        let prefix = format!("data-{idx}");
        let data_files = self
            .write_data_files(batch, &prefix)
            .await
            .map_err(|err| RunError::Other(err.context(format!("entry {idx} write"))))?;

        let tx = Transaction::new(&self.table);
        let tx = tx
            .fast_append()
            .add_data_files(data_files)
            .apply(tx)
            .map_err(|e| RunError::Other(anyhow!("entry {idx} fast_append: {e}")))?;
        self.table = tx
            .commit(&self.catalog)
            .await
            .map_err(|e| RunError::Other(anyhow!("entry {idx} commit: {e}")))?;

        self.record_snapshot()?;
        Ok(())
    }

    /// Run the writer chain (ParquetWriter -> RollingFileWriter -> DataFileWriter)
    /// to turn one RecordBatch into committable DataFiles.
    async fn write_data_files(
        &self,
        batch: RecordBatch,
        prefix: &str,
    ) -> Result<Vec<iceberg::spec::DataFile>> {
        let schema = self.table.metadata().current_schema().clone();
        let loc = DefaultLocationGenerator::new(self.table.metadata())
            .map_err(|e| anyhow!("location generator: {e}"))?;
        let names =
            DefaultFileNameGenerator::new(prefix.to_string(), None, DataFileFormat::Parquet);
        let pw = ParquetWriterBuilder::new(WriterProperties::default(), schema);
        let rolling = RollingFileWriterBuilder::new_with_default_file_size(
            pw,
            self.table.file_io().clone(),
            loc,
            names,
        );
        let mut w = DataFileWriterBuilder::new(rolling)
            .build(None)
            .await
            .map_err(|e| anyhow!("data file writer build: {e}"))?;
        w.write(batch).await.map_err(|e| anyhow!("write: {e}"))?;
        w.close().await.map_err(|e| anyhow!("close: {e}"))
    }

    async fn do_observe(&mut self, idx: usize, e: &Entry) -> Result<(), RunError> {
        let at = parse_at(e).map_err(RunError::Other)?;

        let scan = match &at {
            At::Latest => self.table.scan().build(),
            At::Bind(name) => {
                let snap = *self.binds.get(name).ok_or_else(|| {
                    RunError::Other(anyhow!("entry {idx}: unknown bind {name:?}"))
                })?;
                self.table.scan().snapshot_id(snap).build()
            }
            At::Ordinal(ord) => {
                let snap = self.ordinal_snapshot(*ord).ok_or_else(|| {
                    RunError::Other(anyhow!("entry {idx}: unknown ordinal {ord}"))
                })?;
                self.table.scan().snapshot_id(snap).build()
            }
        }
        .map_err(|e| RunError::Other(anyhow!("entry {idx} scan build: {e}")))?;

        let batches: Vec<RecordBatch> = scan
            .to_arrow()
            .await
            .map_err(|e| RunError::Other(anyhow!("entry {idx} to_arrow: {e}")))?
            .try_collect()
            .await
            .map_err(|e| RunError::Other(anyhow!("entry {idx} collect: {e}")))?;

        let scan_fields = self.table.metadata().current_schema().as_struct().fields();
        let (iceberg_schema, decoded_rows) = decode_scan(&batches, scan_fields, &self.canon_ids)
            .map_err(|err| RunError::Other(err.context(format!("entry {idx} decode"))))?;

        // 'at' echoes the resolved point: a bind name when the observe binds,
        // else the literal target (matches the golden vocabulary).
        let at_value = if let Some(bind) = &e.bind {
            serde_json::Value::String(bind.clone())
        } else {
            match &at {
                At::Latest => serde_json::Value::String("latest".to_string()),
                At::Bind(n) => serde_json::Value::String(n.clone()),
                At::Ordinal(o) => serde_json::Value::Number((*o).into()),
            }
        };

        self.out.observations.push(Observation {
            at: at_value,
            bind: e.bind.clone(),
            iceberg_schema,
            decoded_rows,
        });

        if let Some(bind) = &e.bind
            && let Some(snap) = self.table.metadata().current_snapshot()
        {
            self.binds.insert(bind.clone(), snap.snapshot_id());
        }
        Ok(())
    }

    fn ordinal_snapshot(&self, ordinal: i64) -> Option<i64> {
        self.snapshot_ordinal
            .iter()
            .find_map(|(id, ord)| if *ord == ordinal { Some(*id) } else { None })
    }

    /// Assign the newest snapshot its commit ordinal and record its facts.
    fn record_snapshot(&mut self) -> Result<(), RunError> {
        let snap = self
            .table
            .metadata()
            .current_snapshot()
            .ok_or_else(|| RunError::Other(anyhow!("no current snapshot after commit")))?
            .clone();
        if self.snapshot_ordinal.contains_key(&snap.snapshot_id()) {
            return Ok(());
        }
        let ordinal = self.next_ordinal;
        self.snapshot_ordinal.insert(snap.snapshot_id(), ordinal);
        self.next_ordinal += 1;

        let parent = snap
            .parent_snapshot_id()
            .and_then(|p| self.snapshot_ordinal.get(&p).copied());

        let summary = snap.summary();
        let props = &summary.additional_properties;
        self.out.snapshots.push(SnapshotOut {
            ordinal,
            parent,
            operation: summary.operation.as_str().to_string(),
            summary: summary_out(props),
            delete_files: Vec::new(), // iceberg-rust cannot write delete files (declared gap)
        });
        Ok(())
    }
}

/// Extract the layout-independent aggregate metrics from a snapshot's summary
/// properties. `total-delete-files` is suppressed when zero to match goldens.
fn summary_out(props: &HashMap<String, String>) -> Option<SummaryOut> {
    let get = |k: &str| props.get(k).and_then(|v| v.parse::<i64>().ok());
    let total_records = get("total-records");
    let added_records = get("added-records");
    let deleted_records = get("deleted-records");
    let total_delete_files = get("total-delete-files").filter(|n| *n > 0);
    if total_records.is_none()
        && added_records.is_none()
        && deleted_records.is_none()
        && total_delete_files.is_none()
    {
        return None;
    }
    Some(SummaryOut {
        total_records,
        added_records,
        deleted_records,
        total_delete_files,
    })
}

/// The declared-gap feature name for a delete entry, by kind.
fn delete_feature(e: &Entry) -> String {
    match e.kind.as_deref() {
        Some("equality") => "delete.merge-on-read.equality".to_string(),
        Some("deletion-vector") => "delete.merge-on-read.deletion-vector".to_string(),
        _ => "delete.merge-on-read.position".to_string(),
    }
}

/// Parse the append entry's rows into Row structs.
fn parse_rows(e: &Entry) -> Result<Vec<Row>> {
    let seq = e
        .rows
        .as_ref()
        .and_then(|v| v.as_sequence())
        .ok_or_else(|| anyhow!("append missing rows"))?;
    let mut rows = Vec::with_capacity(seq.len());
    for rv in seq {
        rows.push(parse_row(rv)?);
    }
    Ok(rows)
}

fn parse_row(v: &YamlValue) -> Result<Row> {
    let m = v
        .as_mapping()
        .ok_or_else(|| anyhow!("row must be a mapping"))?;
    let mut row_key = None;
    let mut values = std::collections::BTreeMap::new();
    for (k, val) in m {
        let key = k
            .as_str()
            .ok_or_else(|| anyhow!("row key must be a string"))?;
        if key == ROW_KEY_NAME {
            row_key = Some(
                val.as_str()
                    .ok_or_else(|| anyhow!("__rowkey must be a string"))?
                    .to_string(),
            );
            continue;
        }
        values.insert(key.to_string(), TypedValue::from_yaml(val)?);
    }
    Ok(Row {
        row_key: row_key.ok_or_else(|| anyhow!("row missing __rowkey"))?,
        values,
    })
}

/// Resolve an observe entry's 'at' node.
fn parse_at(e: &Entry) -> Result<At> {
    let at =
        e.at.as_ref()
            .ok_or_else(|| anyhow!("observe missing 'at'"))?;
    match at {
        YamlValue::String(s) if s == "latest" => Ok(At::Latest),
        YamlValue::String(s) => Ok(At::Bind(s.clone())),
        YamlValue::Number(n) => Ok(At::Ordinal(
            n.as_i64()
                .ok_or_else(|| anyhow!("ordinal must be an integer"))?,
        )),
        other => bail!("unsupported 'at' node {other:?}"),
    }
}
